Add instructions on how to use service accounts (#10)
innovocloud/imke-docs: 03.Documentation/01.ServiceAccounts/default.en.md

---

> title: Managing Service Accounts

Reviewer: ```suggestion
title: Managing Kubernetes Service Accounts
```

body_classes: title-center title-h1h2
published: true
---

> Limited access to a kubernetes cluster can be achieved using service accounts and the RBAC feature within kubernetes.

Reviewer: ```suggestion
Limited access to a kubernetes cluster can be achieved using kubernetes service accounts and the RBAC feature within kubernetes.
```

> To achieve this, you will need to create:

Reviewer: ```suggestion
To achieve this, we will need to create:
```

 - service account
 - access role

>  - role binding for the service account to use the access role

Reviewer: ```suggestion
 - role binding for the kubernetes service account to use the access role
```

> All authentication with kubernetes clusters created in IMKE is done using bearer tokens. When you create a new service account a secret will be created for it in the same namespace, and it will be automatically removed when you delete the service account.

Reviewer: Please make sure that you keep line lengths to 80 characters.

> All authentication with kubernetes clusters created in IMKE is done using bearer tokens. When you create a new service account a secret will be created for it in the same namespace, and it will be automatically removed when you delete the service account.

Reviewer: ```suggestion
All authentication with kubernetes clusters created in IMKE is done using bearer tokens. When we create a new kubernetes service account, a secret will be created for it in the same namespace, and it will be automatically removed when you delete the kubernetes service account.
```

> ## Creating a service account

Reviewer: ```suggestion
## Creating a kubernetes service account
```

> To create a service account run the following command and replace `my-serviceaccount` with the name you want to use for the service account:

Reviewer: ```suggestion
To create a kubernetes service account, we run the following command and replace `my-serviceaccount` with the name we want to use for the service account:
```

```bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: ServiceAccount
metadata:
  name: my-serviceaccount
  namespace: my-namespace
EOF
```

> The cluster will then automatically create a new access token with a name like 'my-serviceaccount-token-#####' when the #s are alphanumeric characters. To list the tokens run:

Reviewer: ```suggestion
The cluster will then automatically create a new access token with a name like 'my-serviceaccount-token-#####' when the #s are alphanumeric characters. To list the tokens, we run the following command:
```

```bash
kubectl get secrets --namespace=my-namespace
```

> And then you can print the token you want using the following command and replacing $SECRETNAME with the one that has been created for your service account:

Reviewer: ```suggestion
Now we can print the token we want using the following command and replacing $SECRETNAME with the one that has been created for your service account:
```

```bash
kubectl get secret $SECRETNAME -o jsonpath='{.data.token}' --namespace=my-namespace
```

> Provide the token that has been printed with the name of the service account to a developer or third party to allow them to interact with the cluster.

Reviewer: ```suggestion
We can share the token that has been printed with the name of the kubernetes service account to a developer or third party to allow them to interact with the cluster.
```

> At this point the service account can authenticate with the kubernetes cluster, but is unable to use it do anything. We now need to create a role and a role binding to provide permissions to the service account.

Reviewer: ```suggestion
At this point the kubernetes service account can authenticate with the kubernetes cluster, but is unable to use it do anything, since it does have no rights by default. We need to create a role and a role binding to provide permissions to the kubernetes service account.
```

## Creating authorization permissions

> Kubernetes has two ways of granting access to service accounts, roles and cluster roles. As cluster roles provide access to all namespaces it is recommended that you don't use them unless you have to. We will provide examples on how to use the namespace restricted roles here.

Reviewer: ```suggestion
Kubernetes has two ways of granting access to service accounts: name space restricted roles and cluster roles. As cluster roles provide access to all namespaces we don't use them unless we have to. We will provide examples on how to use the namespace restricted roles in this guide.
```

> Roles provide an explicit whitelisting of permissions for users (both human and service account) and when a user has multiple roles they can do anything that is granted by any of the roles.

Reviewer: ```suggestion
Roles explicitly whitelist permissions for users (both human and service account). When a user has multiple roles, they can do anything that is granted by any of their roles.
```

> To create a role to allow a user to read information about pods in the namespace 'my-namespace' you could use the following command:

Reviewer: ```suggestion
To create a role to allow a user to read information about pods in the namespace 'my-namespace', we use the following command:
```

```bash
kubectl apply -f - <<EOF
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  namespace: my-namespace
  name: pod-reader
rules:
- apiGroups: [""] # "" indicates the core API group
  resources: ["pods"]
  verbs: ["get", "watch", "list"]
EOF
```

> And then to allow the service account we created earlier to use that role, you would use the following command to create a role binding:

Reviewer: ```suggestion
To grant the service account we created earlier to use this role, we use the following command to create a _role binding_:
```

```bash
kubectl create rolebinding read-pods \
  --role=pod-reader \
  --serviceaccount=my-namespace:my-serviceaccount \
  --namespace=my-namespace
```

For a full list of resources run 
```bash
kubectl api-resources

> ```

Reviewer: Please add an empty line here.

And for most resources the available verbs are:
 - get
 - list
 - watch
 - create
 - edit
 - update
 - delete
 - exec

### More information

> The official kubernetes documentation on [controlling access](https://kubernetes.io/docs/reference/access-authn-authz/controlling-access/) and [using roles and role bindings in RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) provides more details

Reviewer: Please add a final section summarizing what we learned in this guide.
